---
output:
  pdf_document:
    number_sections: yes
    toc_depth: 3
    toc: yes
header-includes: \usepackage{float} \floatplacement{figure}{H} \floatplacement{table}{H}
geometry: "left=4cm,right=4cm,top=4cm,bottom=4cm"
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

\begin{itemize}
  \item 3 Different methods currently
  \item 1 NN (hard to interpret), 1 Bayesian (Cell-Segmentations), 1 density based (Segmentation-free)
  \item Use structure of SSAM but improve on methods in each section
\end{itemize}

\section{Introduction}
\begin{itemize}
    \item Mention the larger project, what is this study going to be used for?
    \item 
    \item Mention SSAM, Bayzor, Spage2Vec
    \item 
\end{itemize}

\begin{itemize}
  \item DATA:
  \item I generate data by GMM, variance for x and y for each cluster are taken as 2 samples from uniform dist with parameter min = 3, max = 12
  \item I decide on clusters (250 st)
  \item Number of genes per cell is set by max(30, (area of 90% of theoretical data)*2 + random normal sample with mean 0, sd = 10) to get some variation.
  \item Estimate where we have density peaks by starting at true mean, look at closest neighbors density estimation, second neighbor, third and so on. When we have K times in a row that the density goes down, we found our peak.
\end{itemize}

\section{Method and Theory}

\textcolor{red}{Should I mention SSAM? And then in each step say e.g. they use gaussian kernel with fixt bandwidth but we want an adaptive density estimation so we do...?}
Structure of the method:
\begin{enumerate}
    \item Get genes from tissues as coordinates in 2-D (similar to GMM)
    \item Pointwise density estimations
    \item SSAM do KDE. If bandwidth too large we might get shortcircuits in density estimate for KDE (e.g. in osmFish data at the far right of the bottom part where we have like stairs we might add points from upper stair to the density of a point from the lower stair). Knn is more adaptive.
    \begin{itemize}
        \item Tried one of the most known parameter free estimations (Kraskov et. al).
        \begin{itemize}
             \item Cons: We don't have normalized density estimation or a density function. We use Rank. The rank only depends on the distance to the k:th nearest neighbor
        \item Generalization: Use one of two estimates that make use of all neighbors $1, 2, ..., k$.
        \end{itemize}
        \item Method 1: $\rho_1(i) = 1/\sum_{j = 1}^k d_{ij}$, where $d_{ij}$ is the euclidean distance from point $i$ to point $j$ and $k$ is the number of nearest neighbors. We must fix $k$
        \textcolor{red}{How do we decide on k?}
        \item Method 2: $\rho_2(i) = d_i/vol(V)$ which is the stationary distribution in a similarity graph. Here $d_i = \sum_{j=1}^n \omega_{ij}$ where $\omega_{ij} = \frac{1}{d_{ij}^2 + 1}$ and $vol(V)$ is the sum of $d_i$. We only use ranks so $vol(V)$ does not contribute. 
    \end{itemize}
    \item Some down sampling or local maximum finder \textcolor{red}{How do we find the local maximums? Depends on normalization process.}
    \begin{itemize}
                \item Compared estimated mode and true mode from generated data. Check how well density estimators preserve spatial information of cluster modes I check how many neighbors away from the true cluster mode the estimated mode is.
                \item Figure: Distribution of k from previous point shows peak close to 0 decaying quickly (good).
                \item Further checks: Statistic from SASNE paper.
                \item Method 1 seems to perform better (very slightly)
    \end{itemize}
    \item \textcolor{red}{Check distribution of "True Mode" - "Estimated Mode", i.e. a plot around origo. Are the Estimated Modes biased? I.e. do they always lie to the left of the true mode. Or above? How about variance?}
    \item \textcolor{red}{How does the gaussians generated look like? Does it look good? Bias? Variance?}
    \item Normalization of gene count in maximum
    \begin{itemize}
        \item Standard is sctransform.
        \item Cons: It makes a lot of assumption that do not necessarily hold. e.g. linearity between gene count and sequencing depth (might hold in sc-analysis but not in-situ since and same $r$ in negative binomial for every cell (probably not true).
        \item Assumption that each local maximum vector is the same as a cell. \textcolor{red}{Can we really assume this?}
        \item Sequencing depth and gene count have correlation in sc-sequencing which needs to be normalized. For in-situ we do not do sequencing. \textcolor{red}{Do we need to normalize for sequencing depth? Should we normalize for something else that affects gene count for in-situ samples instead?}
        \item \textcolor{red}{We need another method for this}
    \end{itemize}
    \item Clustering
    \begin{itemize}
        \item \textcolor{red}{Some density based clustering method with soft clustering}
        \item Based on fast search paper.
        \item rho = our density estimations
        \item generalize cutoff point. Look at rho*delta^a and some constant.
        \item Pick the constant where there is a plateue in the graph (bottom left of my picture from meeting)
        \item Try different k in the KNN part. Pick a k where we find a plateue. 
        \item When we have our cluster centers, check if any center has one of its k nearest neighbor in another cluster. If that is the case, k is too large.
    \end{itemize}
    
\end{enumerate}

\bigskip

Notes: 
\begin{itemize}
    \item We have several types of genes. When doing density estimation and finding local maximums we bulk them all together. Afterwards they should be separate.
    \item Pros: Less issue with sparse areas which could be an issue when we do pointwise density estimations.
    \item Cons: Could this implement bias by favoring cells containing genes with high count? \textcolor{red}{Does all cells have same total gene count? Probably not. Then maybe we miss some local max in favor or cells with high gene count}.
\end{itemize}

\section{Results}

\begin{itemize}
  \item Estimated modes and compared with true modes (on different GMM data)
  \item When I checked the distribution of the distance from the true mode to the estimated one, both knn methods had early peaks and the went down. Kde had the same but then another peak at 3sd (the border of the epsilon neighborhood I checked around each true mu). Therefor I redid the procedure with 2sd instead. The same happened at the border of the 2sd epsilon neighborhood instead now
  \item Checked how far it was from the true mode that got bad estimate to the closest other true mode. This was always around the border of epsilon neighborhood. I.e. we take a gene from another cluster as our estimate
\end{itemize}

\begin{itemize}
  \item I tried with different shape clusters aswell now. This gave good results for KNN. KDE got some estimated modes very far away. 
\end{itemize}


\begin{itemize}
    \item Test entire pipeline on GMM generated data.
    \begin{itemize}
        \item How many clusters can we find?
        \item Try generate data from GMM but with different types of genes mimicing real cells. How well is the transcriptomics preserved by the model?
        \item Test the method on benchmark data e.g. osmFish
        \item \textcolor{red}{Test on SciLifeLab data?}
        \item \textcolor{orange}{Interpret and explain why results are as they are!!}
    \end{itemize}
\end{itemize}



# References

(sctransform): Hafemeister, C., Satija, R. Normalization and variance stabilization of single-cell RNA-seq data using regularized negative binomial regression. Genome Biol \textbf{20}, 296 (2019). https://doi.org/10.1186/s13059-019-1874-1

(Spage2Vec): Partel, G. & Wahlby, C. Spage2vec: Unsupervised representation of localized spatial gene expression signatures. FEBS J, \textbf{288}, 1859â€“187 (2021). https://doi.org/10.1111/febs.15572

(Bayzor): Petukhov, V., Xu, R.J., Soldatov, R.A. et al. Cell segmentation in imaging-based spatial transcriptomics. Nat Biotechnol (2021). https://doi.org/10.1038/s41587-021-01044-w

(SSAM): Park, J., Choi, W., Tiesmeyer, S. et al. Cell segmentation-free inference of cell types from in situ transcriptomics data. Nat Commun \textbf{12}, 3545 (2021). https://doi.org/10.1038/s41467-021-23807-4
